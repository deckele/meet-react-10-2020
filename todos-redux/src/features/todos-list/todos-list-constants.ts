export const todosConstants = {
  TODO_CHANGE: "TODO_CHANGE",
  TODO_DELETE: "TODO_DELETE",
  TODO_CREATE: "TODO_CREATE",
} as const;
