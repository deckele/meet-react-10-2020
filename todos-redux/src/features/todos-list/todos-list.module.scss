.todosList {
  list-style: none;
  padding: 0;
}
