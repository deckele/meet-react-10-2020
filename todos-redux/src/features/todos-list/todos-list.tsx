import React from "react";
import { FiltersMap, ITodo } from "../../contracts";
import { useThemeContext } from "../../theme-context";
import { applyFilters } from "../../utils/utils";
import { TodosListItem } from "./todos-list-item/todos-list-item";
import styles from "./todos-list.module.scss";

interface TodosListProps {
  todos: ITodo[];
  filtersMap: FiltersMap;
  todoCreate: () => void;
  todoChange: <TProp extends keyof ITodo>(
    id: string,
    todoProp: TProp,
    value: ITodo[TProp]
  ) => void;
  todoDelete: (id: string) => void;
}
export function TodosList({
  todos,
  filtersMap,
  todoCreate,
  todoChange,
  todoDelete,
}: TodosListProps) {
  const { theme, toggleTheme } = useThemeContext();
  const filterFunctions = Object.values(filtersMap);
  const filteredTodos = applyFilters(todos, filterFunctions);

  return (
    <div style={{ color: theme.primary, backgroundColor: theme.background }}>
      <button onClick={toggleTheme}>Change Theme</button>
      <h4>My Todos</h4>
      <button className={"StyledButton"} onClick={todoCreate}>
        + New Todo
      </button>
      <ul className={styles.todosList}>
        {filteredTodos.map((todo) => (
          <TodosListItem
            key={todo.id}
            {...todo}
            onChange={todoChange}
            onDelete={todoDelete}
          />
        ))}
      </ul>
    </div>
  );
}
