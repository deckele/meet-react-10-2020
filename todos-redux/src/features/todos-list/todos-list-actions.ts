import { ITodo } from "../../contracts";
import { todosConstants } from "./todos-list-constants";

type TodosActionsMap = typeof todosActions;
export type TodosActions = ReturnType<TodosActionsMap[keyof TodosActionsMap]>;
export const todosActions = {
  todoChange<TProp extends keyof ITodo>(
    id: string,
    todoProp: TProp,
    value: ITodo[TProp]
  ) {
    return {
      type: todosConstants.TODO_CHANGE,
      payload: { id, todoProp, value },
    };
  },
  todoCreate() {
    return {
      type: todosConstants.TODO_CREATE,
    };
  },
  todoDelete(id: string) {
    return {
      type: todosConstants.TODO_DELETE,
      payload: { id },
    };
  },
};
