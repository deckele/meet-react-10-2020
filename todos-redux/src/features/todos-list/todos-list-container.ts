import { Dispatch } from "react";
import { connect } from "react-redux";
import { AppState, ITodo } from "../../contracts";
import { TodosList } from "./todos-list";
import { todosActions } from "./todos-list-actions";

function mapStateToProps(state: AppState) {
  return {
    todos: state.todos,
    filtersMap: state.filters,
  };
}
function mapDispatchToProps(dispatch: Dispatch<any>) {
  return {
    todoCreate: () => dispatch(todosActions.todoCreate()),
    todoChange: <TProp extends keyof ITodo>(
      id: string,
      todoProp: TProp,
      value: ITodo[TProp]
    ) => dispatch(todosActions.todoChange(id, todoProp, value)),
    todoDelete: (id: string) => dispatch(todosActions.todoDelete(id)),
  };
}
export const TodosListContainer = connect(
  mapStateToProps,
  mapDispatchToProps
)(TodosList);
