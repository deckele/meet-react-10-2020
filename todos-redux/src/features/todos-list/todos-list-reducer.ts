import { ITodo } from "../../contracts";
import { todosFactory } from "../../utils/utils";
import { TodosActions } from "./todos-list-actions";
import { todosConstants } from "./todos-list-constants";

export type TodosState = ITodo[];

const initialState: TodosState = [
  todosFactory.createTodo("Learn ReactJs", "blue"),
  todosFactory.createTodo("Order from 10Bis", "orange"),
];

export function todosReducer(
  state = initialState,
  action: TodosActions
): TodosState {
  switch (action.type) {
    case todosConstants.TODO_CHANGE: {
      const { id, todoProp, value } = action.payload;
      const todoIndex = state.findIndex((todo) => todo.id === id);
      const changedTodo = { ...state[todoIndex], [todoProp]: value };
      return [
        ...state.slice(0, todoIndex),
        changedTodo,
        ...state.slice(todoIndex + 1),
      ];
    }
    case todosConstants.TODO_CREATE: {
      const newTodo: ITodo = todosFactory.createTodo();
      return [...state, newTodo];
    }
    case todosConstants.TODO_DELETE: {
      const { id } = action.payload;
      return state.filter((todo) => todo.id !== id);
    }
    default: {
      return state;
    }
  }
}
