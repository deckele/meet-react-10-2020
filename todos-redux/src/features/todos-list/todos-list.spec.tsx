import { render, screen } from "@testing-library/react";
import React from "react";
import { TodosList } from "./todos-list";
describe("todos-list tests:", () => {
  it("should render correctly", () => {
    const { asFragment } = render(
      <TodosList
        filtersMap={{}}
        todos={[]}
        todoChange={() => {}}
        todoCreate={() => {}}
        todoDelete={() => {}}
      />
    );
    expect(asFragment()).toMatchSnapshot();
  });
  it("should show todo from todos input", () => {
    const { asFragment } = render(
      <TodosList
        filtersMap={{}}
        todos={[
          { id: "1", done: false, description: "yo my dudes", color: "black" },
        ]}
        todoChange={() => {}}
        todoCreate={() => {}}
        todoDelete={() => {}}
      />
    );
    expect(asFragment()).toMatchSnapshot();
    expect(screen.getByText("Black")).toBeInTheDocument();
    expect(screen.getByDisplayValue("yo my dudes")).toBeInTheDocument();
    expect(screen.queryByDisplayValue("yo my dudes1")).not.toBeInTheDocument();
    screen
      .findByDisplayValue("yo my dudes")
      .then((res) => expect(res).toBeInTheDocument());
  });
});
