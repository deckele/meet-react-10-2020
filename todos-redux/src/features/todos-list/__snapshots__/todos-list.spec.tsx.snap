// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`todos-list tests: should render correctly 1`] = `
<DocumentFragment>
  <div
    style="color: black; background-color: white;"
  >
    <button>
      Change Theme
    </button>
    <h4>
      My Todos
    </h4>
    <button
      class="StyledButton"
    >
      + New Todo
    </button>
    <ul
      class="todosList"
    />
  </div>
</DocumentFragment>
`;

exports[`todos-list tests: should show todo from todos input 1`] = `
<DocumentFragment>
  <div
    style="color: black; background-color: white;"
  >
    <button>
      Change Theme
    </button>
    <h4>
      My Todos
    </h4>
    <button
      class="StyledButton"
    >
      + New Todo
    </button>
    <ul
      class="todosList"
    >
      <li
        class="todosListItem"
      >
        <input
          type="checkbox"
        />
        <input
          class=""
          placeholder="Task description..."
          type="text"
          value="yo my dudes"
        />
        <select
          name="color-picker"
          style="color: black; font-weight: bolder; text-shadow: black 0px 0px 1px;"
        >
          <option
            value="red"
          >
            Red
          </option>
          <option
            value="green"
          >
            Green
          </option>
          <option
            value="blue"
          >
            Blue
          </option>
          <option
            value="cyan"
          >
            Cyan
          </option>
          <option
            value="yellow"
          >
            Yellow
          </option>
          <option
            value="purple"
          >
            Purple
          </option>
          <option
            value="orange"
          >
            Orange
          </option>
          <option
            value="black"
          >
            Black
          </option>
        </select>
        <button>
          X
        </button>
      </li>
    </ul>
  </div>
</DocumentFragment>
`;
